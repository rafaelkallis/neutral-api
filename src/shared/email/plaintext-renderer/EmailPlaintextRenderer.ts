import {
  RoleCtaModel,
  ProjectCtaModel,
  CtaModelWithFirstName,
  CtaModel,
} from 'shared/email/manager/EmailManager';

/**
 *
 */
export abstract class EmailPlaintextRenderer {
  /**
   *
   */
  public abstract renderLoginEmailPlaintext(
    model: CtaModelWithFirstName,
  ): string;

  /**
   *
   */
  public abstract renderSignupEmailPlaintext(model: CtaModel): string;

  /**
   *
   */
  public abstract renderEmailChangeEmailPlaintext(
    model: CtaModelWithFirstName,
  ): string;

  /**
   *
   */
  public abstract renderNewAssignmentEmailPlaintext(
    model: RoleCtaModel,
  ): string;

  /**
   *
   */
  public abstract renderInvitedUserNewAssignmentEmailPlaintext(
    model: RoleCtaModel,
  ): string;

  public abstract renderPeerReviewRequestedEmailPlaintext(
    model: ProjectCtaModel,
  ): string;

  public abstract renderManagerReviewRequestedEmailPlaintext(
    model: ProjectCtaModel,
  ): string;

  public abstract renderProjectFinishedEmailPlaintext(
    model: ProjectCtaModel,
  ): string;
}
