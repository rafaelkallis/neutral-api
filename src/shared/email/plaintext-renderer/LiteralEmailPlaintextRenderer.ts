import { Injectable } from '@nestjs/common';
import { EmailPlaintextRenderer } from 'shared/email/plaintext-renderer/EmailPlaintextRenderer';
import {
  CtaModel,
  ProjectCtaModel,
  RoleCtaModel,
  CtaModelWithFirstName,
} from 'shared/email/manager/EmailManager';

/**
 * Literal Email Plaintext Renderer
 */
@Injectable()
export class LiteralEmailPlaintextRenderer extends EmailPlaintextRenderer {
  public renderLoginEmailPlaintext(model: CtaModelWithFirstName): string {
    return this.renderCtaPlaintext({
      ctaContent: `You have requested a login.`,
      ctaActionLabel: 'Login',
      model,
    });
  }

  /**
   *
   */
  public renderSignupEmailPlaintext(model: CtaModel): string {
    return this.renderCtaPlaintext({
      ctaContent: `You have requested a signup.`,
      ctaActionLabel: 'Signup',
      model: { ...model, firstName: 'there' },
    });
  }

  /**
   *
   */
  public renderEmailChangeEmailPlaintext(model: CtaModelWithFirstName): string {
    return this.renderCtaPlaintext({
      ctaContent: `You have requested an email change.`,
      ctaActionLabel: 'Confirm Email',
      model,
    });
  }

  /**
   *
   */
  public renderNewAssignmentEmailPlaintext(model: RoleCtaModel): string {
    const roleToken = model.roleTitle
      ? `the role of "${model.roleTitle}"`
      : `a role`;
    return this.renderCtaPlaintext({
      ctaContent: `You are assigned ${roleToken} in "${model.projectTitle}".`,
      ctaActionLabel: 'See Assignment',
      model,
    });
  }

  /**
   *
   */
  public renderInvitedUserNewAssignmentEmailPlaintext(
    model: RoleCtaModel,
  ): string {
    const roleToken = model.roleTitle
      ? `the role of "${model.roleTitle}"`
      : `a role`;
    return this.renderCtaPlaintext({
      ctaContent: `You are assigned ${roleToken} in "${model.projectTitle}".`,
      ctaActionLabel: 'Get Started',
      model,
    });
  }

  public renderPeerReviewRequestedEmailPlaintext(
    model: ProjectCtaModel,
  ): string {
    return this.renderCtaPlaintext({
      ctaContent: `You are requested to submit a peer-review in "${model.projectTitle}".`,
      ctaActionLabel: 'Submit Peer Review',
      model,
    });
  }

  public renderManagerReviewRequestedEmailPlaintext(
    model: ProjectCtaModel,
  ): string {
    return this.renderCtaPlaintext({
      ctaContent: `All peer-reviews have been submitted in "${model.projectTitle}" and you are requested to submit a manager-review.`,
      ctaActionLabel: 'Submit Manager Review',
      model,
    });
  }

  public renderProjectFinishedEmailPlaintext(model: ProjectCtaModel): string {
    return this.renderCtaPlaintext({
      ctaContent: `Project "${model.projectTitle}" is finished and the results are ready.`,
      ctaActionLabel: 'Check Results',
      model,
    });
  }

  private renderCtaPlaintext(context: RenderCtaPlaintextContext): string {
    return `
      Hi ${context.model.firstName},

      ${context.ctaContent}

      >> ${context.ctaActionLabel}
      ${context.model.ctaActionUrl}

      - Team Covee
    `;
  }
}

interface RenderCtaPlaintextContext {
  ctaContent: string;
  ctaActionLabel: string;
  model: CtaModelWithFirstName;
}
