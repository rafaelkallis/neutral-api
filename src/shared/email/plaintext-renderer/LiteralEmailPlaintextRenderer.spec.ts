import { LiteralEmailPlaintextRenderer } from 'shared/email/plaintext-renderer/LiteralEmailPlaintextRenderer';
import {
  InvitedUserNewAssignmentModel,
  NewAssignmentModel,
} from '../manager/EmailManager';
import { UnitTestScenario } from 'test/UnitTestScenario';

describe(LiteralEmailPlaintextRenderer.name, () => {
  let scenario: UnitTestScenario<LiteralEmailPlaintextRenderer>;
  let literalEmailPlaintextRenderer: LiteralEmailPlaintextRenderer;

  beforeEach(async () => {
    scenario = await UnitTestScenario.builder(
      LiteralEmailPlaintextRenderer,
    ).build();
    literalEmailPlaintextRenderer = scenario.subject;
  });

  it('should render login text', () => {
    const text = literalEmailPlaintextRenderer.renderLoginEmailPlaintext(
      'https://example.com/login',
    );
    expect(text).toMatchSnapshot();
  });

  it('should render signup text', () => {
    const text = literalEmailPlaintextRenderer.renderSignupEmailPlaintext(
      'https://example.com/signup',
    );
    expect(text).toMatchSnapshot();
  });

  it('should render email change text', () => {
    const text = literalEmailPlaintextRenderer.renderEmailChangeEmailPlaintext(
      'https://example.com/email-change',
    );
    expect(text).toMatchSnapshot();
  });

  it('should render new assignment text', () => {
    const model: NewAssignmentModel = {
      projectTitle: 'Neutron Collider',
      roleTitle: 'Particle Scientist',
      projectUrl: 'http://example.com',
    };
    const text = literalEmailPlaintextRenderer.renderNewAssignmentEmailPlaintext(
      model,
    );
    expect(text).toMatchSnapshot();
  });

  it('should render unregistered user new assignment text', () => {
    const model: InvitedUserNewAssignmentModel = {
      projectTitle: 'Neutron Collider',
      roleTitle: 'Particle Scientist',
      signupMagicLink: 'http://example.com/signup',
    };
    const text = literalEmailPlaintextRenderer.renderInvitedUserNewAssignmentEmailPlaintext(
      model,
    );
    expect(text).toMatchSnapshot();
  });
});
