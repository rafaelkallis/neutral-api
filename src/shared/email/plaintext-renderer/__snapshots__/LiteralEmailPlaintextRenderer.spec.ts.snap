// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiteralEmailPlaintextRenderer should render email change text 1`] = `
"
      Hi Ricky,

      You have requested an email change.

      >> Confirm Email
      https://example.com/email-change

      - Team Covee
    "
`;

exports[`LiteralEmailPlaintextRenderer should render login text 1`] = `
"
      Hi Julian,

      You have requested a login.

      >> Login
      https://example.com/login

      - Team Covee
    "
`;

exports[`LiteralEmailPlaintextRenderer should render manager review requested text 1`] = `
"
      Hi Julian,

      All peer-reviews have been submitted in \\"My Project\\" and you are requested to submit a manager-review.

      >> Submit Manager Review
      http://example.com

      - Team Covee
    "
`;

exports[`LiteralEmailPlaintextRenderer should render new assignment text 1`] = `
"
      Hi Randy,

      You are assigned the role of \\"Particle Scientist\\" in \\"Neutron Collider\\".

      >> See Assignment
      http://example.com

      - Team Covee
    "
`;

exports[`LiteralEmailPlaintextRenderer should render peer review requested text 1`] = `
"
      Hi Cyrus,

      You are requested to submit a peer-review in \\"My Project\\".

      >> Submit Peer Review
      http://example.com

      - Team Covee
    "
`;

exports[`LiteralEmailPlaintextRenderer should render project finished text 1`] = `
"
      Hi Ricky,

      Project \\"My Project\\" is finished and the results are ready.

      >> Check Results
      http://example.com

      - Team Covee
    "
`;

exports[`LiteralEmailPlaintextRenderer should render signup text 1`] = `
"
      Hi there,

      You have requested a signup.

      >> Signup
      https://example.com/signup

      - Team Covee
    "
`;

exports[`LiteralEmailPlaintextRenderer should render unregistered user new assignment text 1`] = `
"
      Hi Bubbles,

      You are assigned the role of \\"Particle Scientist\\" in \\"Neutron Collider\\".

      >> Get Started
      http://example.com/signup

      - Team Covee
    "
`;
